core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -100,16 +99,18 @@ public NodePath(Path path, Environment environment) throws IOException {
 
         /**
          * Resolves the given shards directory against this NodePath
+         * ${data.paths}/nodes/{node.id}/indices/{index.uuid}/{shard.id}
          */
         public Path resolve(ShardId shardId) {
-            return resolve(shardId.getIndex()).resolve(Integer.toString(shardId.id()));
+            return resolve(shardId.getIndex().getUUID()).resolve(Integer.toString(shardId.id()));
         }
 
         /**
-         * Resolves the given indexes directory against this NodePath
+         * Resolves index directory against this NodePath
+         * ${data.paths}/nodes/{node.id}/indices/{index.uuid}
          */
-        public Path resolve(Index index) {
-            return indicesPath.resolve(index.getName());
+        public Path resolve(String indexUUID) {
```

Comment:
can always use the Index class to refer to an index? we should get in this habit so we always have both name and uuid available.